### PR TITLE
Address legion package build issue on macOS with apple-clang-12

### LINKED ISSUE #21261: Installation issue: legion on macos
### Steps to reproduce the issue

```console
$ spack install legion
...
==> Installing legion-20.03.0-ht5mlq3kb4zk65j3qdhefwzxxeqlikyu
==> No binary for legion-20.03.0-ht5mlq3kb4zk65j3qdhefwzxxeqlikyu found: installing from source
/Users/ptb/gitrepos/spack/lib/spack/spack/architecture.py:203: UserWarning: microarchitecture specific optimizations are not supported yet on mixed compiler toolchains [check apple-clang@11.0.3 for further details]
  warnings.warn(msg.format(compiler))
==> Using cached archive: /Users/ptb/gitrepos/spack/var/spack/cache/_source-cache/archive/ae/ae5feedb5ed9f357b56424b9d73cea4f224a61e291e022556f796d1ff24d1244.tar.gz
==> legion: Executing phase: 'cmake'
==> legion: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j12'

4 errors found in build log:
     634        /Users/ptb/gitrepos/spack/opt/spack/darwin-catalina-skylake/apple-clang-11.0.3/gasnet-2020.3.0-eb4avhu4wcxla3mqn6bu5hu
            ec5it6gge/lib/libgasnet-mpi-par.a(gasnet_tools.o)
     635        /Users/ptb/gitrepos/spack/opt/spack/darwin-catalina-skylake/apple-clang-11.0.3/gasnet-2020.3.0-eb4avhu4wcxla3mqn6bu5hu
            ec5it6gge/lib/libgasnet_tools-par.a(gasnet_tools.o)
     636    duplicate symbol '_gasnett_linkconfig_idiotcheck_RELEASE_MAJOR_2020' in:
     637        /Users/ptb/gitrepos/spack/opt/spack/darwin-catalina-skylake/apple-clang-11.0.3/gasnet-2020.3.0-eb4avhu4wcxla3mqn6bu5hu
            ec5it6gge/lib/libgasnet-mpi-par.a(gasnet_tools.o)
     638        /Users/ptb/gitrepos/spack/opt/spack/darwin-catalina-skylake/apple-clang-11.0.3/gasnet-2020.3.0-eb4avhu4wcxla3mqn6bu5hu
            ec5it6gge/lib/libgasnet_tools-par.a(gasnet_tools.o)
     639    ld: 128 duplicate symbols for architecture x86_64
  >> 640    clang: error: linker command failed with exit code 1 (use -v to see invocation)
  >> 641    make[2]: *** [lib/librealm.1.dylib] Error 1
  >> 642    make[1]: *** [runtime/CMakeFiles/RealmRuntime.dir/all] Error 2
  >> 643    make: *** [all] Error 2
```

### Information on your system

```console
$ spack debug report
* **Spack:** 0.16.0-908-85ea52acd
* **Python:** 3.9.1
* **Platform:** darwin-catalina-skylake
* **Concretizer:** original
```

### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/5868839/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/5868834/spack-build-env.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [X] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

spack maintainer: @streichler

### PR DESCRIPTION
**Pull request by pmccormick**
Work-around for apple-clang-12 build issue -- clang does not appear to default to the documented C++ standard as stated in the clang man page.  This fix introduces a variant for C++ standard support/selection (defaulting to `-std=c++11` to match Legion's current CI configurations).

Should resolve #21261.

### CODE REVIEW

File: var/spack/repos/builtin/packages/legion/package.py [hunk 1]
```diff
@@ -62,6 +62,13 @@ class Legion(CMakePackage):
     depends_on('papi', when='+papi')
     depends_on('zlib', when='+zlib')
 
+    # A C++ standard variant to work-around some odd behaviors with apple-clang
+    # but this might be helpful for other use cases down the road.  Legion's
+    # current development policy is C++11 or greater so we capture that aspect
+    # here.
+    cpp_stds = ["11", "14", "17", "20"]
+    variant('c++std', default="11", values=cpp_stds, multi=False)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/legion/package.py]: For consistency with other packages, I recommend:

```suggestion
    variant('cxxstd', default="11", values=cpp_stds, multi=False)
```

File: var/spack/repos/builtin/packages/legion/package.py [hunk 2]
```diff
@@ -201,7 +208,10 @@ def validate_gasnet_root(value):
     def cmake_args(self):
         spec = self.spec
         cmake_cxx_flags = []
-        options = []
+        from_variant = self.define_from_variant
+        options = [
+            from_variant("CMAKE_CXX_STANDARD", "c++std")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/legion/package.py]: See reason above.

```suggestion
            from_variant("CMAKE_CXX_STANDARD", "cxxstd")
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/legion/package.py (hunk 2 above)
Comment (pmccormick) [reply var/spack/repos/builtin/packages/legion/package.py]: Thanks @tldahlgren!  Done.

**Review by tldahlgren — APPROVED** [review]

**Comment by pbrady** [comment]
It worked!  Thanks @pmccormick

**Pull request merged**